### ISSUE TITLE
Remove duplicate and obsolete bbcodes and possibly modify others.

### ISSUE THREAD

**Issue by Rupurudu** [SimpleMachines/SMF#2885]
While answering an user for a list of all bbcodes for the board, I found many duplicate and obsolete bbcodes in SMF

Worser, these bbcodes are are also included on SMF 2.1. And more horrible changes are made. :(

`[acronym]` = It's a duplicate of `[abbr]` and marked as obsolete on HTML5. on SMF 2.1, it uses <abbr title="$1"> instead of removing the tag. :(

`[br]` = This shouldn't be in SMF. We automatically replace newlines with br tag. Furthermore this also brokes our code too. Newline limitation on signatures are not applied if this code used as newline.

`[black]`, `[white]`, `[red]`, `[green]`, `[blue]` = Another useless and duplicate tag. `[color]` tag should be used instead of this tags. Ironically they somehow set a css value on SMF 2.1.

`[tt]` = Exactly the same as color ones `[font=monospace]` should be used instead, and on SMF 2.1 it's replaced with a CSS value.

`[ltr]`, `[rtl]`, `[bdo]` = Either remove `[bdo]` and keep `[ltr]`, `[rtl]` or remove `[ltr]`, `[rtl]` and keep `[bdo]` instead.
Suggestion: use `[ltr]` and `[rtl]`

`[email]`, `[ftp]`, `[iurl]`, `[url]` = 4 tags, all makes links with different protocols. Why? No one knows. 

`[flash]` = noembed tag is not an w3c standard and they recommend to not use it. See: http://www.w3.org/TR/html5/obsolete.html#obsolete 
Also, play="true" loop="true" quality="high" are not needed, imo.

`[flash]`, `[img]` = validate section should validate ftp:// and ftps:// too.

`[quote]` = Remove `<div class="quotefooter"></div>` It's not used on SMF 2.0 and SMF 2.1. It's also not used on custom themes. 

`[li]` = Should work like `[*]` if the end tag is not properly closed.
`[@]`, `[*]` = Duplicate
`[O]`, `[0]`, `[o]` = Another duplicate
`[+]`, `[x]`, `[#]` = Wow. Another one.

`[chrissy]`, `[kissy]` = ???

**Comment by FearThe1337** [comment]
```
[email], [ftp], [iurl], [url] = 4 tags, all makes links with different protocols. Why? No one knows.
```

I assume this has to do with the fact that these work differently. the [iurl vs url](http://wiki.simplemachines.org/smf/Iurl) is nicely explained here:

>   iurl: These links will normally open in the same browser window
>   For links that will open in a new window by default, use the url tag.

I would assume, I have not verified that ftp and email are for that same purpose. If a ftp tag is used without ftp://, ftp:// automatically gets added.

```
[tt] = Exactly the same as color ones [font=monospace] should be used instead, and on SMF 2.1 it's replaced with a CSS value.
```

I'm not sure how often this is used, but I assume more often then other fonts? In this case a shortcut seems handy?

```
[flash], [img] = validate section should validate ftp:// and ftps:// too.
```

I am not too sure about if that is a good idea or not. In all fairness ftp and ftps are not made to be used in a webbrowser.

Also regarding a lot of your mentioned bbcodes, when searching "[name] bbcode" on google I found that results vary a lot, (different forum names being found), I assume these duplicate tags are there to make bbcode from other forum software work on smf too.

**Comment by MissAllSunday** [comment]
I'll be in favor of removing those too.

chrissy and kissy are easter eggs and will remain there as a sign of respect to the previous devs.

**Comment by illori** [comment]
if you remove them, will you not break posts that used them from when they did exist?

**Comment by FearThe1337** [comment]
@illori 
As far as I can see posts are stored in html, so they should not break.

**Comment by illori** [comment]
it is parsed as html, that is as far as i know only because SMF knows what the bbc should be.

the database still has the bbc code present

```
test<br />[acronym=test]test[/acronym]
```

the code you post is what is in the database but a blank line is replaced with the br.

**Comment by FearThe1337** [comment]
Hmm correct, only < br > is placed in the database. Which means it would break old posts.

**Comment by margarett** [comment]
I would strongly suggest that the number of posts that will break due to this are minimal. Who uses those BBCs anyway?! So yes, please remove the bloatware...

chrissy and kissy... Really? :open_mouth:

**Comment by illori** [comment]
we dont know across all SMF forums how often those are used and if a lot of posts would break with their removal.

**Comment by Oldiesmann** [comment]
The "chrissy" and "kissy" tags are Easter eggs that have been in there since the beginning. I don't care what we do with those either way.

As far as the rest - I would really prefer not to remove them because, as Illori said, we don't know what is or isn't used. If there are problems with HTML validation (such as the "noembed" thing), we may be able to rewrite the HTML to better conform with modern web standards.

I also don't think img should support the FTP protocol - I have never seen this protocol used in this manner in all the years (15+) I've been on the web and FTP links themselves aren't even all that common these days.

**Comment by MissAllSunday** [comment]
Remove them and provide an official mod for those who want them back...

**Comment by Oldiesmann** [comment]
What benefit do we gain from removing them?

**Comment by MissAllSunday** [comment]
Some marginal cycle saving I guess.

Its not so much about actual, tangible benefits its more about getting rid of some pretty old/unused BB code. I mean, SMF 2.1 still shows the glow and marquee icons in the editor...

People has this preconceive notion of SMF being old and this kind of obsolete BBC codes still present on 2.1 surely helps in keeping that stereotype alive...

**Comment by margarett** [comment]
^^ Amen to that :)

**Comment by bperry921** [comment]
if they are not broken, why remove them?

**Comment by bperry921** [comment]
SMF is old, that's a fact, and one of the more versatile as well, these type of features have made it that way...

**Comment by MissAllSunday** [comment]
Why is there such a (nostalgic?) sudden interest for all those BBC codes?

Unless there is plenty of facts and clear evidence that  BBC tags such as glow and marquee are still actively used by at least 30% of the forums powered by SMF then I see absolutely no reason why we should keep them.

People still (for reasons that scape my own sanity) uses those BBc tags, thats true, however, I am willing to bet that the people still using those tags doesn't represent 5% of the total SMF installations and quite frankly, I've been generous saying 5% as it should be close to 1%

So, for those 5% that still uses those BBC tags and assuming all of them are going to migrate to SMF 2.1, the best course of actions is to provide them with an official modification.

**Comment by bperry921** [comment]
So what you are saying, is that we take away features as a way to gain a wider audience then? IMO we should focus on expanding what's new about SMF

**Comment by MissAllSunday** [comment]
> So what you are saying, is that we take away features as a way to gain a wider audience then? IMO we should focus on expanding what's new about SMF

Not even close, sorry.

**Comment by Sesquipedalian** [comment]
The ability to render obsolete BBCodes must remain, or else backwards compatibility will be lost for no good reason. The ability to use obsolete BBCodes in _new_ posts, on the other hand, could be disabled by default without causing any harm.

**Comment by MissAllSunday** [comment]
No backward compatibility would be lost, these BBC tags are hardly used by anyone, you cannot lose something you don't even know its there.

For the last time, this is NOT about features or compatibility (no one, I repeat NO ONE will ever complain about us removing these BBC tags, I guarantee it), this is purely a cosmetic stuff, these BBC are old and obsolete and us having them been prominently displayed on our editor feels and looks awful, it make us look old and outdated and for no good reason.

Again, unless we have solid evidence and stats of people actually using those BBC tags (which we don't have BTW) I see absolutely no reason why we should still show them, let alone support them.

For the, maybe 5 users who still uses those BBC tags, an official mod should be released.

This is going to be the very, very, very, very first thing I will do for 3.0.....

**Comment by Sesquipedalian** [comment]
I agree that the buttons for obsolete BBCodes should be removed from the editor. But the ability to render the obsolete BBCodes can and should be left in place when the buttons are removed.

For example, there's no need to show an editor button for [tt]. It's a useless BBCode, for the reasons explained above. But if a user decided to use it in a post written with an older version of SMF, we should still render their post the way they intended.

**Comment by MissAllSunday** [comment]
Thing is, why do we bother to keep something that only a handful of users are ever going to use?  Thats my point, if these BBC tags were still largely used by SMF users then yes, keep them up!  but the reality is that the percentage of SMF users who uses those BBC is pretty small, I don't recall reading any support issue or anyone saying they use those BBC tags.

If an admin has a forum with messages that uses those BBC thats where the official mod will chime in. I'm not advocating for completely removing these features, (although we should...) I say remove them from SMF and offer an official mod for those interested on keep that functionality.

The very low percentage of users who still uses this or has messages with these BBC tags is not enough to justify keeping them on the software, instead, offer an official mod for them, it just seems more logical.

**Comment by Sesquipedalian** [comment]
Because keeping the ability to render the obsolete BBC harms nothing. If we want SMF to look more modern, removing the obsolete BBC buttons from the editor does that job all by itself. Ripping out the code to process and render obsolete BBC wouldn’t actually do anything to help further with that. It’d just be more work for everyone.

In fact, if you want obsolete BBC to go the way of the dodo, it actually makes MORE sense to leave the rendering code in place and just remove the buttons. If existing posts still work even when the buttons are gone, there is very little incentive to install a mod. But if one has to install a mod just to get the old posts to render correctly, now the darn buttons are back, too. 

In other words, if we render the obsolete BBCodes in existing posts but remove the ability to create them easily in new posts, they’ll sail off quietly into the sunset. But if we forcibly terminate them, they’ll keep crawling out of their graves like undead abominations to infect people’s forums with their horror.

**Comment by MissAllSunday** [comment]
I give up, SMF can stay in 1997. Heck, I propose we add a Java applet BBC and a flash intro too! maybe we need to include some totally awesome DHTML code.

This should be at the bottom of each page:

![Best view with IE](http://i.imgur.com/03HV88g.gif)

best viewed with IE 6.0

**Comment by live627** [comment]
hehehehe

**Comment by Rupurudu** [comment]
@MissAllSunday @Sesquipedalian You both understand me wrong. All of these bbcodes have modern replacements already on SMF itself. Nothing is lost. We can upgrade the database, and all of them should render normal.

`[acronym]` can be replaced with `[abbr]`
`[br]` can be replaced with newline character
`[black]`, `[white]`, `[red]`, `[green]`, `[blue]` can be replaced with `[color=black]`, `[color=white]`, `[color=red]`, `[color=green]`, `[color=blue]`
and so on...

**Comment by illori** [comment]
but can there be a fool proof way to convert from one bbc to another without breaking the posts?

**Comment by Rupurudu** [comment]
Yes, while we upgrade the forum we also run a upgrade.sql file. We can upgrade the old bbcodes here.

**Comment by illori** [comment]
i know that upgrade_*.sql is run but i dont think it is 100% reliable to edit the posts and bbc. also you are not considering those that may use portal mods or other mods that you can use bbc in that may use these bbc that can never be corrected when you run the upgrade script.

**Comment by live627** [comment]
Also, forums converted from other software may be using them.

**Comment by Rupurudu** [comment]
mods and converters should handle it out themselves, like every change we made.

Also, honestly I don't think there are not many people who uses these tags, since they are not shown on the editor itself.

**Comment by Rupurudu** [comment]
I started to work on this. Here's a few changes to first post:

`[b]` `[i]` `[u]` `[s]` = `<b>` `<i>` `<u>` `<s>` are re-added on HTML5. Fallbacks are no longer neccesary.

`[iurl]` is needed for `[anchor]`, so I'm not removing it.

However, I have some questions:

`[email]`, `[ftp]`, `[url]` = Some admins will disable email tag to protect user info. How can I add multiple permission settings for 1 bbcode? For example: "`[url]` for mailto:" "`[url]` for ftp://" etc. 

Also, should we support other protocols for `[url]`? Here are few good and bad examples:

> Good ones:
> steam:// Used on Steam client.
> magnet: Used on BitTorrent client.
> irc: Used on IRC.
> git:// Used on Git client.
> 
> Bad ones:
> javascript: XSS :frowning: 

---

Also, `[glow]` is a duplicate of `[shadow]`, and both are using ms-filter styled attributes. We should modernize them and support text-shadow styled attributes.

**Comment by XinYenFon** [comment]
My vote goes to support those well known (good ones), maybe with extra (origin://).

**Comment by Rupurudu** [comment]
Why we are removing shadow? Sure, IE only filters needs to be removed but we can implement CSS text-shadow instead.

**Comment by colinschoen** [comment]
How often do you think text-shadow is used? This might be a good thing to pull a small bit of the SMF community on at sm.org

**Comment by XinYenFon** [comment]
Well considering the discussions and work on this issue stopped, I suggest to close this for this version and plan better in next version.

**Comment by sbulen** [comment]
I agree.    No clear end-state defined or agreed upon.
